fail2ban/tests/clientreadertestcase.py

```diff
@@ -649,8 +649,9 @@ def testReadStockJailConfForceEnabled(self):
 
 			# and we know even some of them by heart
 			for j in ['sshd', 'recidive']:
-				# by default we have 'auto' backend ATM
-				self.assertTrue(['add', j, 'auto'] in comm_commands)
+				# by default we have 'auto' backend ATM, but some distributions can overwrite it, 
+				# (e.g. fedora default is 'systemd') therefore let check it without backend...
+				self.assertTrue([True for i in comm_commands if len(i) == 3 and i[0:2] == ['add', j]])
```

Comment:
On Wed, 27 Jul 2016, Serg G. Brester wrote:
>  >                       # and we know even some of them by heart
>  >                       for j in ['sshd', 'recidive']:
>  > -                             # by default we have 'auto' backend ATM
>  > -                             self.assertTrue(['add', j, 'auto'] in comm_commands)
>  > +                             # by default we have 'auto' backend ATM, but some distributions can overwrite it,
>  > +                             # (e.g. fedora default is 'systemd') therefore let check it without backend...
>  > +                             self.assertTrue([True for i in comm_commands if len(i) == 3 and i[0:2] == ['add', j]])

>      performance difference is imho negligible

>    Collection of 2000 negligible non performant differences ea. 1ms, makes
>    the whole thing 2 seconds longer...
>    Since then I've optimized our test-cases in 0.10, I respect it very
>    strong, in particular by such easy-going things, that everyone can read
>    (I'm about readability)...

>      if you want the same -- make 2nd one a generator

>    nope, this generator is larger, because contains more entries...

>    Secondly that will be slower and slower with each extending of jail, regex
>    of jail or filter, etc...

nope 

1. your construct is not a generator at all -- it is a list
comprehension, it will go through **all** the entries

2. mine is a generator, it will stop going through the entries as soon
as it finds a match, it doesn't store anything along the way.

$> cat /tmp/1.py
def visit(i):
   print "visited ", i
   return i


print 3 in (visit(x) for x in range(10))

$> python /tmp/1.py
visited  0
visited  1
visited  2
visited  3
True


but interestingly you are right for the case whenever desired entry is the last
in the list, but then very much off if it is in the beginning ;) 

regardless of those -- it takes miniscule time to run, and I have
spent already more time altogether on comprehending original code and
then arguing to simplify it than probably all the runs of this test in
the next year ;)  so I would still vote for more readable code.

$> cat /tmp/time_check.py
from timeit import timeit

j = "ssh"

global f1

def f1(comm_commands):
    return bool([True for i in comm_commands if len(i) == 3 and i[0:2] == ['add', j]])

def f2(comm_commands):
    return ["add", j] in (x[:2] for x in comm_commands)


comm_commands = [["add", str(x)] for x in range(1000)]
comm_commands_last = comm_commands + [['add', j]]
comm_commands_first = [['add', j]] + comm_commands


kw = dict(
    setup="from __main__ import f1, f2, comm_commands_first, comm_commands_last",
    number=1000
)

print "checks"
print "f1: %s, %s" % (f1(comm_commands_first), f1(comm_commands_last))
print "f2: %s, %s" % (f2(comm_commands_first), f2(comm_commands_last))

if __name__ == '__main__':

    print "Last entry"
    for f in ('f1', 'f2'):
        print "%s: %.7f" % (f, timeit("%s(comm_commands_last)" % (f,), **kw)/kw['number'])

    print "First entry"
    for f in ('f1', 'f2'):
        print "%s: %.7f" % (f, timeit("%s(comm_commands_first)" % (f,), **kw)/kw['number'])


$> python /tmp/time_check.py
checks
f1: False, False
f2: True, True
Last entry
f1: 0.0000471
f2: 0.0001337
First entry
f1: 0.0000476
f2: 0.0000006